modules/db_virtual/dbase.c

```diff
@@ -512,3 +512,180 @@ int db_virtual_insert_update(const db_con_t* _h, const db_key_t* _k,
+				 * then things are gonna be messed up if continuing */          \
+				LM_ERR("failover call failed rc:%d\n", rc);                     \
+				/* set local can not use flag*/                                 \
+				handle->flags &= NOT_CAN_USE;                                   \
+                                                                                \
+				/* close connection*/                                           \
+				set_update_flags(CURRCON(__ah), p);                             \
+                                                                                \
+				f->close(handle->con);                                          \
+				/* if failed before placing the fd in reactor                   \
+				 * we keep on */                                                \
+				if ((--__ah->cons_rem) == 0) {                                  \
+					LM_ERR("All databases failed!! No hope for you!\n");        \
+					return -1;                                                  \
+				}                                                               \
+                                                                                \
+				/* try next*/                                                   \
+				rc = -1;                                                        \
+				CURRCON(__ah)  = (CURRCON(__ah)+1)%p->size;                     \
+			} else {                                                            \
+				if (_resume_f)                                                  \
+					async_status = ASYNC_CHANGE_FD;                             \
+				set_update_flags(CURRCON(__ah), p);                             \
+				return rc;                                                      \
+			}                                                                   \
+		} else {                                                                \
+			LM_DBG("flags2 = %i\n", p->con_list[CURRCON(__ah)].flags);          \
+			if ((--__ah->cons_rem) == 0) {                                      \
+				LM_ERR("All databases failed!! No hope for you!\n");            \
+				return -1;                                                      \
+			}                                                                   \
+                                                                                \
+			/* try next*/                                                       \
+			rc = -1;                                                            \
+			CURRCON(__ah)  = (CURRCON(__ah)+1)%p->size;                         \
+		}                                                                       \
+		LM_DBG("curent_con = %i\n", CURRCON(__ah));                             \
+	} while ((__ah)->cons_rem); /* should never exit here */                    \
+                                                                                \
+	return rc;                                                                  \
+}while (0);
```

Comment:
[This](https://sourcemaking.com/files/sm/images/spagett.jpg) is all I can say about the runaway macros in `db_virtual`.